### DISCUSSION TITLE
Customizing span kind in Netty server with auto-instrumentation

### DISCUSSION THREAD

**Discussion by orangain** [open-telemetry/opentelemetry-java-instrumentation#8139] [Q&A] [answered]
I have built a web application using Netty server (and Ktor) and set up Open Telemetry's auto-instrumentation using javaagent. In this case, Netty server instrumentation automatically creates spans with the `SERVER` Kind, but I want to configure it as follows:

- If the request path starts with `/from-pubsub/`: Kind should be `CONSUMER`
- Otherwise: Kind should be `SERVER`

How can I configure this?

Alternatively, if it is possible to change the Kind of an automatically created span midway, that would also be fine, but I think it might be difficult based on the following issue:

https://github.com/open-telemetry/opentelemetry-java/issues/1793

----

Versions:
- Ktor Server 2.2.4
- Netty: 4.1.86
- Open Telemetry: 1.24.0

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @orangain ,

I can think of 2 ways to achieve that:
1. You could probably use the collector's [transform processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor) to change the span kind for the given route
2. Or, you'd have to manually instrument your server code, and turn off the ktor javaagent instrumentation. This'd also require making some changes in the ktor library instrumentation to allow modifying the `SpanKindExtractor`

**Reply by orangain** [reply]
@mateuszrzeszutek  Thank you for your answer! Sorry for the late reply.

Way 1 is an attractive solution, but I am using a fully managed Google Cloud Run environment and cannot use sidecar containers, i.e. collectors. I am sending traces directly from the application to the APM vendor endpoint.

For this reason, I would consider using way 2. As you say, implementing it in the application would require code changes in Ktor instrumentation, which is a concern though.

**Reply by mateuszrzeszutek** [reply]
> For this reason, I would consider using way 2. As you say, implementing it in the application would require code changes in Ktor instrumentation, which is a concern though.

Just to clarify: I'm not saying that you should fork the Ktor instrumentation; this kind of change would be a good addition to this project.

**Reply by trask** [reply]
> this kind of change would be a good addition to this project

I'm not sure I understand the proposal here, can you outline (very) briefly?

**Reply by mateuszrzeszutek** [reply]
Having an option to pass a `Function<SpanKindExtractor, SpanKindExtractor>` to the Ktor library instrumentation. There are some library instrumentations that allow customizing the span name or status in a similar way; I think we could do the same for the span kind.

**Reply by orangain** [reply]
@mateuszrzeszutek Thank you for the clarification! I will send a pull request for that change.

**Reply by mateuszrzeszutek** [reply]
Thanks!

**Reply by orangain** [reply]
I've created https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8255.
